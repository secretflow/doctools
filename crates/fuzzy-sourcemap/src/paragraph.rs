use swc_core::common::{SourceFile, Span};

use crate::{
  one_indexed,
  whitespace::{indentation_of, is_empty_or_whitespace},
};

pub fn find_paragraph(src: &SourceFile, line_number: usize) -> Option<Span> {
  let line_number = one_indexed!(line_number);

  let line_number = {
    let mut current_line = line_number;
    loop {
      let Some(text) = src.get_line(current_line) else {
        break current_line;
      };
      if !is_empty_or_whitespace(&*text) {
        break current_line;
      } else {
        let Some(next_line) = current_line.checked_sub(1) else {
          break current_line;
        };
        current_line = next_line;
      }
    }
  };

  let line_content = src.get_line(line_number);

  let Some(line) = line_content else {
    return None;
  };

  let indent = indentation_of(&*line);

  fn find_nearest_indented(
    src: &SourceFile,
    start: usize,
    direction: i8,
    indentation: &str,
  ) -> usize {
    let mut current_line = start;

    let get_next_line = if direction < 0 {
      |u: usize, d: i8| u.checked_sub(d.abs() as usize)
    } else {
      |u: usize, d: i8| u.checked_add(d as usize)
    };

    loop {
      let Some(next_line) = get_next_line(current_line, direction) else {
        return current_line;
      };
      let Some(next_line_text) = src.get_line(next_line) else {
        return current_line;
      };
      let next_line_text = &*next_line_text;
      if !indentation_of(next_line_text).starts_with(indentation)
        || is_empty_or_whitespace(next_line_text)
      {
        return current_line;
      } else {
        current_line = next_line;
      }
    }
  }

  let lower = find_nearest_indented(src, line_number, -1, indent);
  let upper = find_nearest_indented(src, line_number, 1, indent);

  let (lower, _) = src.line_bounds(lower);
  let (_, upper) = src.line_bounds(upper);

  Some((lower, upper).into())
}
