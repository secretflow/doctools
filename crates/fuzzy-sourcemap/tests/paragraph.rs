use fuzzy_sourcemap::paragraph::find_paragraph;

mod utils;

#[must_use]
fn test_case(case: &str, line: usize) -> Option<String> {
  let sources = utils::get_fixture();
  let source = utils::get_fixture_file(&sources, case);
  let para = find_paragraph(&*source, line);
  match para {
    Some(span) => Some(
      sources
        .with_snippet_of_span(span, |text| String::from(text))
        .unwrap(),
    ),
    None => None,
  }
}

#[test]
fn test_find_paragraph_markdown_013() {
  insta::assert_snapshot!(test_case("markdown.md", 13).unwrap());
}

#[test]
fn test_find_paragraph_markdown_017() {
  insta::assert_snapshot!(test_case("markdown.md", 17).unwrap());
}

#[test]
fn test_find_paragraph_markdown_031() {
  insta::assert_snapshot!(test_case("markdown.md", 31).unwrap());
}

#[test]
fn test_find_paragraph_markdown_054() {
  insta::assert_snapshot!(test_case("markdown.md", 54).unwrap());
}

#[test]
fn test_find_paragraph_markdown_140() {
  insta::assert_snapshot!(test_case("markdown.md", 140).unwrap());
}

#[test]
fn test_find_paragraph_markdown_138() {
  insta::assert_snapshot!(test_case("markdown.md", 138).unwrap());
}

#[test]
fn test_find_paragraph_markdown_182() {
  insta::assert_snapshot!(test_case("markdown.md", 182).unwrap());
}

#[test]
fn test_find_paragraph_markdown_318() {
  insta::assert_snapshot!(test_case("markdown.md", 318).unwrap());
}

#[test]
fn test_find_paragraph_markdown_341() {
  insta::assert_snapshot!(test_case("markdown.md", 341).unwrap());
}

#[test]
fn test_find_paragraph_markdown_352() {
  insta::assert_snapshot!(test_case("markdown.md", 352).unwrap());
}

#[test]
fn test_find_paragraph_markdown_374() {
  insta::assert_snapshot!(test_case("markdown.md", 374).unwrap());
}

#[test]
fn test_find_paragraph_markdown_393() {
  insta::assert_snapshot!(test_case("markdown.md", 393).unwrap());
}

#[test]
fn test_find_paragraph_markdown_410() {
  assert_eq!(test_case("markdown.md", 410), None);
}

#[test]
fn test_find_paragraph_empty() {
  assert_eq!(test_case("empty.md", 1), None);
}
