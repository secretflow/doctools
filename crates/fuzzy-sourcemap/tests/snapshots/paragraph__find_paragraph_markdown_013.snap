---
source: crates/fuzzy-sourcemap/tests/paragraph.rs
expression: "test_case(\"markdown.md\", 13).unwrap()"
---
Markdown is a plain text format for writing structured documents,
based on conventions for indicating formatting in email
and usenet posts. It was developed by John Gruber (with
help from Aaron Swartz) and released in 2004 in the form of a
[syntax description](https://daringfireball.net/projects/markdown/syntax)
and a Perl script (`Markdown.pl`) for converting Markdown to
HTML. In the next decade, dozens of implementations were
developed in many languages. Some extended the original
Markdown syntax with conventions for footnotes, tables, and
other document elements. Some allowed Markdown documents to be
rendered in formats other than HTML. Websites like Reddit,
StackOverflow, and GitHub had millions of people using Markdown.
And Markdown started to be used beyond the web, to author books,
articles, slide shows, letters, and lecture notes.

